mlflow/models/evaluation/base.py

```diff
@@ -151,6 +156,194 @@ def __str__(self):
         return "EvaluationMetric(" + ", ".join(parts) + ")"
 
 
+# NB: we need this function because we cannot modify the signature of
+# a class's __call__ method after the class has been defined.
+# This is also useful to distinguish between the metric signatures with different eval_fn signatures
+def dynamically_generate_eval_metric(eval_fn, strict_signature=False):
+    """
+    Dynamically generate a GenAIEvaluationMetric class that can be used to evaluate the metric
+    on the given input data. The generated class is callable with a __call__ method that
+    takes the arguments specified in the signature of the eval_fn function.
+
+    Args:
+        eval_fn: the evaluation function of the EvaluationMetric.
+        strict_signature: (Optional) Whether the metric follows a strict signature, if True then
+            the eval_fn must follow below signature:
+
+                .. code-block:: python
+
+                    def eval_fn(
+                        predictions: "pd.Series",
+                        metrics: Dict[str, MetricValue],
+                        inputs: "pd.Series",
+                        *args,
+                    ) -> MetricValue:
+                        pass
+
+            When generating a metric from `make_genai_metric`, this should be set to True.
+            Default to False.
+
+    Returns:
+        A dynamically generated callable CallableEvaluationMetric class.
+    """
+    from mlflow.metrics.base import MetricValue
+
+    if strict_signature:
+        allowed_kwargs_names = [
+            param_name
+            for param_name in inspect.signature(eval_fn).parameters.keys()
+            if param_name not in ["predictions", "metrics", "inputs"]
+        ]
+
+        def genai_call_method(
+            self,
+            *,
+            predictions: Union[pd.Series, str, list],
+            inputs: Union[pd.Series, str, list],
+            metrics: Optional[Dict[str, MetricValue]] = None,
+            return_only_scores: bool = False,
+            **kwargs,
+        ) -> Union[MetricValue, List[str], List[float]]:
```

Comment:
When does the return value become `List[str]`?

Reply:
MetricValue.scores has type `Union[List[str], List[float]]`, is it not accurate?

Reply:
It seems the string type was introduced in https://github.com/mlflow/mlflow/commit/a1db4c34be6d3f93878ed326baa478d67e3984d4. Tbh I couldn't read why we need it from PR description/comments.... but it is as it is, let's keep both.